### ISSUE TITLE
Cannot reconnect to persisted config due to verify-x509-name

### ISSUE THREAD

**Issue by fleischie** [OpenVPN/openvpn3-linux#90]
Hello,

I am using `openvpn3` on Arch Linux installed from AUR but without DCO (was producing errors).
```
karl@archCarbon ~ $ openvpn3 version
OpenVPN 3/Linux git:makepkg:713b35e908489579 (openvpn3)
OpenVPN core 3.git:HEAD:b47c72b4 linux x86_64 64-bit
Copyright (C) 2012-2020 OpenVPN Inc. All rights reserved.
```

Importing a config and connecting to a session works as expected, but when I import a config with `--persistent` I am not able to re-connect to the stored config.

Looking at the error message it says: "Connection failed: option_error: Invalid verify-x509-name type" hinting at the first space character of the `profile.verify-x509-name` attribute of the configuration json.

E.g.:
- "C=DE, L=CN, O=Company Name" -> "Invalid verify-x509-name type: L=OE",
- "C=DE,L=CN,O=Company Name" -> "Invalid verify-x509-name type: Name".

Which might indicate, that the name is not correctly checked for certain characters. Might also be for non-ascii chars but I haven't checked that thouroughly.

Is this even correct here, or should this rather copy this issue to the openvpn3-core repo? Let me know if you need additional info and/or have hints/notes/etc.

**Comment by fleischie** [comment]
Thank you for taking the time and putting in the work to investigate this so thoroughly. 🙇

Do you still need me to set something up, test something, or provide further info?

**Comment by dsommers** [comment]
No, I'm all fine now.  I've been able to isolate the problem and can reproduce it at will now.

The problematic code path is related to a similar fix we added for `--static-challenge` in commit ef26761c9d21cccb8ca62a1883787552f745024d.  But that was a hack, and this hack can't be reused for `--verify-x509-name`.  I'm working on a fix in the Core library's [option rendering method](https://github.com/OpenVPN/openvpn3/blob/7fa1c70821d5f51f86a3a156f872200f21c2e89a/openvpn/common/options.hpp#L258), which should also remove the need for the hack added for `--static-challenge`.

**Comment by dsommers** [comment]
This should now be fixed with commit 87570c0a3663507ec339df5bec373a7e4ffed67b.  This does a complete overhaul of the code importing and exporting the JSON data for the persistent storage, which helps removing several other hacks as well.

This approach will hopefully be much better in the end.

**Comment by dsommers** [comment]
And just to give more details in the issue itself.

### What happened
When the configuration file was imported, it was stored correctly in memory inside the Core library's configuration object.  But for the persistent storage, the json_export() used the OptionList::render() method which did not encode the `verify-x509-name` option correctly, as the double quotes were stripped out.  So the JSON data stored on disk looked like this:
        
        "verify-x509-name": "C=DE, L=CN, O=Company Name",

The crucial detail here is that when this is imported back, the value read is `C=DE, L=CN, O=Company Name` and not `"C=DE, L=CN, O=Company Name"`.  So when the OpenVPN 3 Core library later on at connection point parses this option, it sees the `verify-x509-name` with 3 independent values (`C=DE`, `L=CN` and `O=Company Name`) and not as a single, combined value.  That makes the `VerifyX509Name` option parser complain, because the optional second value (referred to as "type" in the docs) can only be `subject` (default when not provided), `name` or `name-prefix`.  The incorrectly parsed option value on the second position is `L=CN`, which does not match a valid type.  So it complains and stops running.

This error does also not appear until after the `openvpn3-service-configmgr` has stopped and reloads the incorrectly formatted JSON file.  If the configmgr service has the initially imported configuration file in memory, it works fine - as that is correctly stored in memory.  It is the restart of the configmgr service which reveals this issue.

### The fix
The fix extends the JSON format a bit and will make use of the same storage model on disk as the OpenVPN 3 Core library uses internally.  Instead handling the complete option value as a single string (which would in some cases require proper quoting and text processing at import), the new JSON storage will make use of arrays instead.

The configuration data is already properly tokenized in memory by the Core library option parser. These tokens are now preserved when exporting the data as JSON - using an array per option, each token is an array element - and the same tokens are restored into memory again when importing the JSON data.

Even though the JSON import parser gets a bit convoluted now, that is to ensure it can read already existing files.  These files will be upgraded to the new format on the first VPN sessoin use or if other profile updates happens.  The various JSON format versions is now also documented in the source code.

**Comment by StuartSwitzman** [comment]
No problem

**Comment by fliespl** [comment]
@dsommers was this fix released? If so - in what versions of OpenVPN Connect.

**Comment by dsommers** [comment]
OpenVPN 3 Linux v18_beta has been released.  This issue should be resolved in that release.

https://www.mail-archive.com/openvpn-devel@lists.sourceforge.net/msg24464.html
